Add Three20 object with version information.

jverkoey opened the pull request:
This adds code-level version information to Three20Core. This can be used to determine which version of the Three20 framework someone is using.

jverkoey committed:
[Core] Add Three20 object with version information.

jverkoey commented:
Related pull request: #514

jverkoey committed:
MERGEME: Add %d for the hotfix version.

diederich commented:
While I like the idea of being able to query the version in code, I'm not a fan of maintaining this manually 'in-code'. Additionally it'd be nice to have a commit hash, too.

I must say I haven't really looked into Xcode/Apple's versioning stuff, e.g. agvtools, VERSION_INFO_FILE (Three20Core_vers.c) etc., but did you think about it?

jverkoey commented:
Definitely open to a better solution for maintaining this file. Do you have experience with any of those versioning tools? I haven't used them myself. If they're too complex for our needs, we could write a small script to set the version numbers programmatically (with a commit hash, as well).

diederich commented:
Nope, sorry, haven't played with them, but pinged tonklon. Guess he'll comment later...

jwang commented:
@jverkoey check this out. maybe something along those lines work as well.
http://www.cimgf.com/2008/04/13/git-and-xcode-a-git-build-number-script/
http://kswizz.com/post/2686511526/git-xcode-versioning

jverkoey commented:
Let's try this mention thing out: @tonklon

jverkoey commented:
For historical purposes, here are the current places where we manage our version info:
- `src/Three20/Headers/Three20.h` - Used by doxygen to generate the api home page and link to the roadmap. We can probably use `src/Three20Core/Headers/Three20.h` now to do this.
- On the Three20.info backend, [doxygen.config](https://github.com/Three20/Backend/blob/master/scripts/doxygen/doxygen.config) must be updated.

jverkoey commented:
@jwang That bundle version script looks handy. I don't think we currently version our bundle.

tonklon commented:
Mentioning works. I'll look into that next week. I think Apple generic version tool can be of help here.

jverkoey commented:
Cool, pardon the mention spam tonklon. Looking forward to hearing back. Thanks for checking in on this.

tonklon commented:
No, agvtool wont help here. It assumes the version is a double.
From agvtool(8):

```
The target of a versioned project will have two global variables generated and linked
into your product.  One is of type double and is simply the CURRENT_PROJECT_VERSION. 
The other is a version string which is formatted to be compatible with what(1). These variables
are available for use in your code.
```

And even if this would not be the case, it is not possible to include the version number into the documentation this way. Doxygen runs the standard C preprocessor before generating documentation, but we would need a #define holding the version for this to work.

So I think a dedicated version header holding a preprocessor definition of the current version would be the best way I can see. This should be accessible by Doxygen for the docs and by the Three20 class.

jverkoey commented:
Is there a specific standard we should follow for defining this version header? Would Three20Version.h suffice?

tonklon commented:
I have no idea at all.
But Three20Version.h looks good to me.

jverkoey commented:
Cool, thanks again tonklon for your help with this.

I'll go ahead and keep the Three20.h/m files added in this commit, but add a Three20Version.h that is the sole place where we define the Three20 version number.

jverkoey committed:
[Core] Add Three20Version.h and unify the locations where version is defined.

Also fleshed out the unit tests for the Three20 object.

jverkoey commented:
Added Three20Version.h and co. This satisfies the criteria of having our version defined only once in the project now.

jverkoey committed:
[Core] Three20Version is now a global extern NSString* defined in Three20Version.m.

Much better than a macro. Changing the library version won't require all dependencies
to recompile.

jverkoey committed:
[Core] Make Three20Version a const pointer so that it cannot be modified.

jverkoey commented:
Are we good to go on this guy?

diederich commented:
yep
